PR: https://github.com/odoo/odoo/pull/104741

From: 75a105f46a13f75eb56a0de80faa2a6e146730aa
From: Gorash
Date: 2023-08-18 07:49:11

Structural Changes: 3.6
Total Changes: 12

[REF] base/all: Update modifier syntax: remove 'states' from fields

These changes are made as a result of simplifying attrs and 'states' in
views. However, they should have remained in a separate commit. When
applying the script making the xml changes (used later for the migration
script), the script checked the definition of the python fields in order
to convert the information into a python expression. Therefore, this
commit is not applied when the script is applied to xml changes.

During this attribute deletion pre-existing errors were found. Part of
the code was using the boolean values of 'states' and another part of
the code was not. The behavior could therefore be different (in cases
where readonly on the field had the same value as the ballan in
'states').

Following the deletion of 'states' and without the application of the
view migration, the js tests (tower) were no longer functional. Tests
using the Form view suffered the same effect. There are few tests that
had to be adapted, including two tests in business accounting (updated
by the accounting team). A test for column_invisible did not work. Test
checking if the test system triggers an error if we try to write on an
invisible field. It turns out that Form was testing on the value of
invisible but not taking into account if the column was invisible. The
test system fix is applied separately because there were a lot of tests
that were incorrect.

Part-of: odoo/odoo#104741

================================= pseudo patch: =================================

--- a/addons/stock_picking_batch/models/stock_picking.py
+++ b/addons/stock_picking_batch/models/stock_picking.py
@@ -65,7 +65,6 @@ class StockPicking(models.Model):
     batch_id = fields.Many2one(
         'stock.picking.batch', string='Batch Transfer',
         check_company=True,
-        states={'done': [('readonly', True)], 'cancel': [('readonly', True)]},
         help='Batch associated to this transfer', index=True, copy=False)
 
     @api.model_create_multi

--- a/addons/stock_picking_batch/models/stock_picking_batch.py
+++ b/addons/stock_picking_batch/models/stock_picking_batch.py
@@ -19,14 +19,13 @@ class StockPickingBatch(models.Model):
         copy=False, required=True, readonly=True)
     user_id = fields.Many2one(
         'res.users', string='Responsible', tracking=True, check_company=True,
-        readonly=True, states={'draft': [('readonly', False)], 'in_progress': [('readonly', False)]})
+        readonly=False)
     company_id = fields.Many2one(
         'res.company', string="Company", required=True, readonly=True,
         index=True, default=lambda self: self.env.company)
     picking_ids = fields.One2many(
-        'stock.picking', 'batch_id', string='Transfers', readonly=True,
+        'stock.picking', 'batch_id', string='Transfers', readonly=False,
         domain="[('id', 'in', allowed_picking_ids)]", check_company=True,
-        states={'draft': [('readonly', False)], 'in_progress': [('readonly', False)]},
         help='List of transfers associated to this batch')
     show_check_availability = fields.Boolean(
         compute='_compute_move_ids',
@@ -42,8 +41,7 @@ class StockPickingBatch(models.Model):
         'stock.move', string="Stock moves", compute='_compute_move_ids')
     move_line_ids = fields.One2many(
         'stock.move.line', string='Stock move lines',
-        compute='_compute_move_ids', inverse='_set_move_line_ids', readonly=True,
-        states={'draft': [('readonly', False)], 'in_progress': [('readonly', False)]})
+        compute='_compute_move_ids', inverse='_set_move_line_ids', readonly=False)
     state = fields.Selection([
         ('draft', 'Draft'),
         ('in_progress', 'In progress'),
@@ -53,12 +51,11 @@ class StockPickingBatch(models.Model):
         copy=False, tracking=True, required=True, readonly=True, index=True)
     picking_type_id = fields.Many2one(
         'stock.picking.type', 'Operation Type', check_company=True, copy=False,
-        readonly=True, index=True, states={'draft': [('readonly', False)]})
+        index=True)
     picking_type_code = fields.Selection(
         related='picking_type_id.code')
     scheduled_date = fields.Datetime(
         'Scheduled Date', copy=False, store=True, readonly=False, compute="_compute_scheduled_date",
-        states={'done': [('readonly', True)], 'cancel': [('readonly', True)]},
         help="""Scheduled date for the transfers to be processed.
               - If manually set then scheduled date for all transfers in batch will automatically update to this date.
               - If not manually changed and transfers are added/removed/updated then this will be their earliest scheduled date
